File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Users/Classes/PlaceholderGenerator/PlaceholderGenerator.m
//
//  PlaceholderGenerator.m
//  LoginComponent
//
//  Created by Andrey Ivanov on 08/06/16.
//  Copyright © 2016 Quickblox. All rights reserved.
//

#import "PlaceholderGenerator.h"

@interface PlaceholderGenerator()

@property (strong, nonatomic) NSCache *cache;
@property (strong, nonatomic) NSArray *colors;

@end

@implementation PlaceholderGenerator

+ (instancetype)instance {
    
    static PlaceholderGenerator *_generator = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        
        _generator = [[PlaceholderGenerator alloc] init];
    });
    
    return _generator;
}

- (instancetype)init {
    
    self = [super init];
    if (self) {
        
        _cache = [[NSCache alloc] init];
        _cache.name = @"QMUserPlaceholer.cache";
        _cache.countLimit = 200;
        
        _colors =
        @[[UIColor colorWithRed:0.9906f green:0.5137f blue:0.033 alpha:1.0f], // #FF9600
          [UIColor colorWithRed:0.267f green:0.859f blue:0.369f alpha:1.0f], // #44DB5E
          [UIColor colorWithRed:0.2781f green:0.7296f blue:0.984f alpha:1.0f], // #54C7FC
          [UIColor colorWithRed:0.9862f green:0.0499f blue:0.2661f alpha:1.0f], // #FF2D55
          [UIColor colorWithRed:0.5304f green:0.0727f blue:0.6196f alpha:1.0f], // #9B2FAE
          [UIColor colorWithRed:0.082f green:0.584f blue:0.533f alpha:1.0f], // #159588
          [UIColor colorWithRed:0 green:0.478f blue:1.0f alpha:1.0f], // #007AFF
          [UIColor colorWithRed:0.804f green:0.855f blue:0.286f alpha:1.0f], // #CDDA49
          [UIColor colorWithRed:0.122f green:0.737f blue:0.823f alpha:1.0f], // #1FBCD2
          [UIColor colorWithRed:0.251f green:0.329f blue:0.698f alpha:1.0f]]; // #4054B2
    }
    
    return self;
}

- (UIColor* _Nonnull)colorForString:(NSString*)string {
    
    unsigned long hashNumber = stringToLong((unsigned char*)[string UTF8String]);
    UIColor* color = self.colors[hashNumber % [self.colors count]];
    return color;
}

unsigned long stringToLong(unsigned char* str) {
    
    unsigned long hash = 5381;
    int c;
    while ((c = *str++)) {
        hash = ((hash << 5) + hash) + c;
    }
    return hash;
}

+ (UIImage *)placeholderWithSize:(CGSize)size
                           title:(NSString *)title {
    
    NSString *key = [NSString stringWithFormat:@"%@ %@", title, NSStringFromCGSize(size)];
    
    UIImage *image = [[PlaceholderGenerator instance].cache objectForKey:key];
    
    if (image) {
        
        return image;
    }
    else {
        
        UIImage *img = [self placeholderWithSize:size
                                           color:[[PlaceholderGenerator instance] colorForString:title]
                                           title:title isOval:YES];
        
        [[PlaceholderGenerator instance].cache setObject:img forKey:key];
        
        return img;
    }
}

+ (UIImage *)placeholderWithSize:(CGSize)size color:(UIColor *)color title:(NSString *)title isOval:(BOOL)isOval {
    
    NSUInteger min = MIN(size.width, size.height);
    CGRect frame = CGRectMake(0, 0, size.width, size.height);
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0);
    
    UIBezierPath *path = nil;
    if (isOval) {
        
        path = [UIBezierPath bezierPathWithOvalInRect:frame];
    }
    else {
        
        path = [UIBezierPath bezierPathWithRect:frame];
    }
    
    [color setFill];
    [path fill];
    
    NSMutableParagraphStyle *paragraphStyle = [[NSMutableParagraphStyle defaultParagraphStyle] mutableCopy];
    paragraphStyle.alignment = NSTextAlignmentCenter;
    
    UIFont *font = [UIFont systemFontOfSize:min / 2.0f];
    UIColor *textColor = [UIColor whiteColor];
    
    NSString *textContent = [[title substringToIndex:1] uppercaseString];
    
    NSDictionary *ovalFontAttributes = @{NSFontAttributeName:font ,
                                         NSForegroundColorAttributeName:textColor,
                                         NSParagraphStyleAttributeName:paragraphStyle};
    
    CGRect rect = [textContent boundingRectWithSize:frame.size
                                            options:NSStringDrawingUsesLineFragmentOrigin
                                         attributes:ovalFontAttributes context:nil];
    
    CGRect textRect = CGRectOffset(frame,
                                   0,
                                   (size.height - rect.size.height) / 2);
    [textContent drawInRect:textRect withAttributes: ovalFontAttributes];
    //Get image
    UIImage *img = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return img;
}

+ (UIImage *)groupPlaceholderWithUsers:(NSArray *)users
                                  size:(NSUInteger)size {
    
    CGSize tSize = (CGSize){size, size};
    
    UIGraphicsBeginImageContextWithOptions(tSize, NO, 0.0);
    
    NSUInteger count = MIN(users.count, 4);
    
    for (NSUInteger i = 0; i < count; i++) {
        
        QBUUser *user = users[i];
        CGRect r = [self rectWithIndex:i size:size count:count];
        
        UIImage *img = [self placeholderWithSize:r.size
                                           color:[PlaceholderGenerator instance].colors[user.ID % 10]
                                           title:user.fullName
                                          isOval:NO];
        
        [img drawInRect:r];
    }
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();

    UIGraphicsBeginImageContextWithOptions(tSize, NO, 0.0);
    
    [[UIColor colorWithRed:0.5 green:0.5 blue:0.5 alpha:0.3] setFill];
    [[UIBezierPath bezierPathWithOvalInRect:CGRectMake(0, 0, size, size)] fill];
    
    CGRect interiorBox = CGRectInset(CGRectMake(0, 0, size, size), 2, 2);
    UIBezierPath *interior = [UIBezierPath bezierPathWithOvalInRect:interiorBox];
    
    [interior addClip];
    [image drawInRect:CGRectMake(0, 0, size, size)];
    
    UIImage *finalImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return finalImage;
}

+ (CGRect)rectWithIndex:(NSUInteger)idx size:(NSUInteger)size count:(NSUInteger)count {
    
    NSUInteger h = count > 2 ? size / 2 : size;
    NSUInteger s = size/2;
    
    switch (idx) {
        case 0:
            
            return CGRectMake(0, 0, s, count < 4 ?  size : size / 2);
            
        case 1:
            
            return CGRectMake(s, 0, s, h);
            
        case 2:
            
            return CGRectMake(count < 4 ? s : 0, s, s, h);
            
        case 3:
            
            return CGRectMake(s, s, s, h);
    }
    
    return CGRectZero;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Users/Classes/PlaceholderGenerator/PlaceholderGenerator.swift
//
//  PlaceholderGenerator.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 17.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox

class PlaceholderGenerator {
    
    static let instance = PlaceholderGenerator()
    
    private lazy var cache: NSCache<AnyObject, AnyObject> = {
        let cache = NSCache<AnyObject, AnyObject>()
        cache.name = "QMUserPlaceholer.cache"
        cache.countLimit = 200
        return cache
    }()
    
    private let colors: [UIColor] = [#colorLiteral(red: 0.9529411793, green: 0.6862745285, blue: 0.1333333403, alpha: 1), #colorLiteral(red: 0.3035047352, green: 0.8693258762, blue: 0.4432001114, alpha: 1), #colorLiteral(red: 0.2588235438, green: 0.7568627596, blue: 0.9686274529, alpha: 1), #colorLiteral(red: 0.8549019694, green: 0.250980407, blue: 0.4784313738, alpha: 1), #colorLiteral(red: 0.02297698334, green: 0.6430568099, blue: 0.603818357, alpha: 1), #colorLiteral(red: 0.5244195461, green: 0.3333674073, blue: 0.9113605022, alpha: 1), #colorLiteral(red: 0, green: 0.5694751143, blue: 1, alpha: 1), #colorLiteral(red: 0.839125216, green: 0.871129334, blue: 0.3547145724, alpha: 1), #colorLiteral(red: 0.09088832885, green: 0.7803853154, blue: 0.8577881455, alpha: 1), #colorLiteral(red: 0.3175504208, green: 0.4197517633, blue: 0.7515394688, alpha: 1)]
    
    class func color(index: Int) -> UIColor {
        return PlaceholderGenerator.instance.color(index: index)
    }
    
    private func color(index: Int) -> UIColor {
        let color = colors[index % 10]
        return color
    }
    
    class func placeholder(size: CGSize, title: String?) -> UIImage {
        let key = title ?? ""
        guard let image = PlaceholderGenerator.instance.cache.object(forKey: key as AnyObject) as? UIImage  else {
            let index = key.count % 10
            let image = placeholder(size: size,
                                    color: PlaceholderGenerator.instance.color(index: index),
                                    title: title,
                                    isOval: true)
            PlaceholderGenerator.instance.cache.setObject(image, forKey: key as AnyObject)
            return image
        }
        return image
    }
    
    class func placeholder(size: CGSize, color: UIColor, title: String?, isOval: Bool) -> UIImage {
        let minSize = min(size.width, size.height)
        let frame = CGRect(x: 0, y: 0, width: size.width, height: size.height)
        
        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)
        
        let path = isOval ? UIBezierPath(ovalIn: frame) : UIBezierPath(rect: frame)
        
        color.setFill()
        path.fill()
        
        let paragraphStyle = NSMutableParagraphStyle()
        paragraphStyle.alignment = .center
        
        let font = UIFont.systemFont(ofSize: minSize / 2.0)
        let textColor = UIColor.white
        let titleString = NSString(string: title ?? "Q")
        
        let textContent = titleString.substring(to: 1).uppercased()
        
        let attributes: [NSAttributedString.Key: Any] = [.font: font,
                                                         .foregroundColor: textColor,
                                                         .paragraphStyle: paragraphStyle]
        
        let rect = textContent.boundingRect(with: frame.size,
                                            options: .usesLineFragmentOrigin,
                                            attributes: attributes,
                                            context: nil)
        
        let textRect = frame.offsetBy(dx: (size.width - rect.width) / 2.0,
                                      dy: (size.height - rect.height) / 2.0)
        
        textContent.draw(in: textRect, withAttributes: attributes)
        //Get image
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return image ?? UIImage()
    }
    
    class func groupPlaceholder(withUsers users: [QBUUser]?, size: Int) -> UIImage? {
        
        let tSize: CGSize? = CGSize(width: size, height: size)
        
        UIGraphicsBeginImageContextWithOptions(tSize ?? CGSize.zero, _: false, _: 0.0)
        
        let count = min((users?.count)!, 4)
        
        for i in 0..<count {
            
            let user = users?[i]
            let r: CGRect = self.rect(withIndex: i, size: size, count: count)
            let indexUser = user?.id ?? 0
            let index: Int = Int(indexUser % 10)
            
            let img: UIImage? = self.placeholder(size: r.size, color: PlaceholderGenerator.instance.colors[index], title: user?.fullName, isOval: true)
            img?.draw(in: r)
        }
        
        let image: UIImage? = UIGraphicsGetImageFromCurrentImageContext()
        
        UIGraphicsEndImageContext()
        
        UIGraphicsBeginImageContextWithOptions(tSize ?? CGSize.zero, _: false, _: 0.0)
        
        UIColor(red: 0.5, green: 0.5, blue: 0.5, alpha: 0.3).setFill()
        UIBezierPath(ovalIn: CGRect(x: 0, y: 0, width: CGFloat(size), height: CGFloat(size))).fill()
        
        let interiorBox = CGRect(x: 0, y: 0, width: CGFloat(size), height: CGFloat(size)).insetBy(dx: 2, dy: 2)
        let interior = UIBezierPath(ovalIn: interiorBox)
        
        interior.addClip()
        image?.draw(in: CGRect(x: 0, y: 0, width: CGFloat(size), height: CGFloat(size)))
        
        let finalImage: UIImage? = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return finalImage
    }
    
    class func rect(withIndex idx: Int, size: Int, count: Int) -> CGRect {
        
        let h: Int = count > 2 ? size / 2 : size
        let s: Int = size / 2
        
        switch idx {
        case 0:
            return CGRect(x: 0, y: 0, width: s, height: count < 4 ? size : size / 2)
        case 1:
            return CGRect(x: s, y: 0, width: s, height: h)
        case 2:
            return CGRect(x: count < 4 ? s : 0, y: s, width: s, height: h)
        case 3:
            return CGRect(x: s, y: s, width: s, height: h)
        default:
            return CGRect.zero
        }
    }
}
